### PR TITLE
Implement DSCInitialization Controller

### PR DESCRIPTION
**Pull request by VaishnaviHire**
Introduces changes defined in #245 

## Description
- [x] Ability to create Namespaces
- [x]  Ability to create namespace specific Network Policies, RoleBindings
- [x] Identifying if Addon Installation
- [x]  Ability to add Addon specific logic
- [x]  Optional Dev field to refer custom manifests 
- [ ]  Deploy Monitoring resources (Partial implementation, paused any changes until further confirmation regarding implementation of Monitoring stack)
- [x] Deploy Dashboard CRDs
- [ ]  Add Logs and Events
- [x] Cleanup resources 
- [x] Ability of Operator to create DSCInitialization CR
- [ ] Add Unit Tests for DSCInitialization Controller
- [ ]  Add ConsoleLink for Managed services

### Test already built image

1. Deploy Operator
```
 operator-sdk run bundle quay.io/vhire/opendatahub-operator-bundle:dev-0.0.1
```
2. Verify `default` DSCInitialization instance is created
3. Verify `opendatahub` namespace is created with networkpolicies and odh specific labels

### Test custom image build
1. Pull in updated manifests
```
make get-manifests -e MANIFEST_REPO=VaishnaviHire -e MANIFEST_RELEASE=rearch_manifests
```
2. Build operator Image
```
 make image -e IMG=<quay.io/username/opendatahub-operator:<TAG>>
```
4. Build Bundle
```
 make bundle-build bundle-push BUNDLE_IMG=quay.io/<username>/opendatahub-operator-bundle:<TAG>
```
5. Run Bundle
```
operator-sdk run bundle -e BUNDLE_IMG=quay.io/<username>/opendatahub-operator-bundle:<TAG>
```
## Issues
Fix Path for local manifest bundle

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: config/manager/kustomization.yaml [hunk 1]
```diff
@@ -12,5 +12,5 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 images:
 - name: controller
-  newName: quay.io/vhire/opendatahub-operator
-  newTag: dev-0.0.1
+  newName: registry.jharmison.com/library/opendatahub-operator
```
Comment (strangiato) [inline config/manager/kustomization.yaml:15]: This is probably left over from testing but I assume this shouldn't be merged.

**Comment by VaishnaviHire** [comment]
From today's sync up -
- update the PR to add reconciliation for Network Policies 
- keep Monitoring resources configurable

Moving to wip to address above comments

File: config/manager/kustomization.yaml (hunk 1 above)
Comment (VaishnaviHire) [reply config/manager/kustomization.yaml:15]: Yes, will revert this change

File: config/crd/kustomization.yaml [hunk 2]
```diff
@@ -2,9 +2,12 @@
 # since it depends on service name and namespace that are out of this kustomize package.
 # It should be run by config/default
 resources:
-- bases/dscinitialization.opendatahub.io_dscinitiatlizations.yaml
 - bases/dscinitialization.opendatahub.io_dscinitializations.yaml
 - bases/datasciencecluster.opendatahub.io_datascienceclusters.yaml
+- bases/odhapplications.dashboard.opendatahub.io_odhapplications.yaml
+- bases/odhdashboardconfigs.opendatahub.io_odhdashboardconfigs.yaml
+- bases/odhdocuments.dashboard.opendatahub.io_odhdocuments.yaml
+- bases/odhquickstarts.console.openshift.io_odhquickstarts.yaml
```
[the remark this answers is not in the document]
Comment (VaishnaviHire) [reply config/crd/kustomization.yaml:10]: This kustomize format is generated by the operator-sdk/kubebuilder when we run `operator-sdk generate manifests`.

You are correct, with respect to having `default` folder. For application manifests, we do check for `kustomization.yaml` file or `default`. So components are expected to have a `default` manifest path.

File: config/manager/kustomization.yaml (hunk 1 above)
Comment (jharmison-redhat) [reply config/manager/kustomization.yaml:15]: Oops - apologies :)

File: controllers/dscinitialization/dscinitialization_controller.go [hunk 3]
```diff
@@ -18,50 +18,57 @@ package dscinitialization
...
-	r.Log = r.Log.WithValues("Request.Namespace", req.Namespace, "Request.Name", req.Name)
-	r.ctx = ctx
-
-	r.Log.Info("Reconciling DSCInitialization.", "DSCInitialization", klog.KRef(req.Namespace, req.Name))
+	r.Log.Info("Reconciling DSCInitialization.", "DSCInitialization", req.Namespace, "Request.Name", req.Name)
 
 	instance := &dsci.DSCInitialization{}
-	err := r.Client.Get(ctx, req.NamespacedName, instance)
+	// Only apply reconcile logic to 'default' instance of DataScienceInitialization
+	err := r.Client.Get(ctx, types.NamespacedName{Name: "default"}, instance)
```
Comment (etirelli) [inline controllers/dscinitialization/dscinitialization_controller.go:66]: very minor nitpick, maybe set a constant for "default" as it is also used in the predicate below?

**Review by etirelli — APPROVED** [review]
LGTM

**Pull request merged**